src/librustc_typeck/check/demand.rs

```diff
@@ -374,6 +383,66 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         None
     }
 
+    /// This function checks if the specified expression is a built-in range literal.
+    /// (See: ``librustc::hir::lowering::LoweringContext::lower_expr()``).
+    fn is_range_literal(&self, expr: &hir::Expr) -> bool {
+        use hir::{Path, QPath, ExprKind, TyKind};
+
+        // We support `::std::ops::Range` and `::core::ops::Range` prefixes
+        let is_range_path = |path: &Path| {
+            let mut segs = path.segments.iter()
+                .map(|seg| seg.ident.as_str());
+
+            if let (Some(root), Some(std_core), Some(ops), Some(range), None) =
+                (segs.next(), segs.next(), segs.next(), segs.next(), segs.next())
+            {
+                // "{{root}}" is the equivalent of `::` prefix in Path
+                root == "{{root}}" && (std_core == "std" || std_core == "core")
+                    && ops == "ops" && range.starts_with("Range")
+            } else {
+                false
+            }
+        };
+
+        let is_range_literal = |span: &Span| {
+            // Tell if expression span snippet doesn't look like an explicit
+            // Range struct or `new()` call.  This is to allow inferring
```

Comment:
"Inferring" has a very specific meaning, so I'd tend to avoid it in favour of something like:
```rust
// Check whether a span corresponding to a range is a range literal,
// rather than an explicit struct or `new()` call.
```